lib/semantics/expression.cc

```diff
@@ -1514,17 +1514,97 @@ auto ExpressionAnalyzer::AnalyzeProcedureComponentRef(
   return std::nullopt;
 }
 
+// Can actual be argument associated with dummy?
+static bool CheckCompatibleArgument(bool isElemental,
+    const ActualArgument &actual, const characteristics::DummyArgument &dummy) {
+  return std::visit(
+      common::visitors{
+          [&](const characteristics::DummyDataObject &x) {
+            characteristics::TypeAndShape dummyTypeAndShape{x.type};
+            if (!isElemental && actual.Rank() != dummyTypeAndShape.Rank()) {
+              return false;
+            } else if (auto actualType{actual.GetType()}) {
+              return dummyTypeAndShape.type().IsTkCompatibleWith(*actualType);
+            } else {
+              return false;
+            }
+          },
+          [&](const characteristics::DummyProcedure &) {
+            const auto *expr{actual.UnwrapExpr()};
+            return expr && IsProcedurePointer(*expr);
+          },
+          [&](const characteristics::AlternateReturn &) {
+            return actual.isAlternateReturn;
+          },
+      },
+      dummy.u);
+}
+
+// Are the actual arguments compatible with the dummy arguments of procedure?
+static bool CheckCompatibleArguments(
+    const characteristics::Procedure &procedure,
+    const ActualArguments &actuals) {
+  bool isElemental{procedure.IsElemental()};
+  const auto &dummies{procedure.dummyArguments};
+  CHECK(dummies.size() == actuals.size());
+  for (std::size_t i{0}; i < dummies.size(); ++i) {
+    const characteristics::DummyArgument &dummy{dummies[i]};
+    const std::optional<ActualArgument> &actual{actuals[i]};
+    if (!actual || !CheckCompatibleArgument(isElemental, *actual, dummy)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+const Symbol *ExpressionAnalyzer::ResolveGeneric(
+    const Symbol &symbol, ActualArguments &actuals) {
+  struct {
+    const Symbol *elemental{nullptr};  // matching elemental specific
+    const Symbol *nonElemental{nullptr};  // matching non-elemental specific
+  } result;
+  const auto &details{symbol.get<semantics::GenericDetails>()};
+  for (const Symbol *specific : details.specificProcs()) {
+    if (std::optional<characteristics::Procedure> procedure{
+            characteristics::Procedure::Characterize(
+                ProcedureDesignator{*specific}, context_.intrinsics())}) {
+      parser::Messages buffer;
+      parser::ContextualMessages messages{
+          context_.foldingContext().messages().at(), &buffer};
+      FoldingContext localContext{context_.foldingContext(), messages};
+      ActualArguments localActuals{actuals};
+      if (CheckExplicitInterface(*procedure, localActuals, localContext) &&
+          CheckCompatibleArguments(*procedure, localActuals)) {
+        auto *ptr{procedure->IsElemental() ? &result.elemental
+                                           : &result.nonElemental};
+        CHECK(*ptr == nullptr);
```

Comment:
In an error recovery situation where a generic interface has non-distinguishable specific procedures, this `CHECK` could fail.

Reply:
I could probably check for the error flag on the symbol to avoid this.
But the logic would be simpler if I don't bother with this consistency check.
Then I can return as soon as a compatible non-elemental specific is found.